src/wifi_dpp.rs

```diff
@@ -0,0 +1,451 @@
+//! Wi-Fi Easy Connect (DPP) support
+//!
+//! # Example
+//!
+//! Note that to use this feature, you must add CONFIG_WPA_DPP_SUPPORT=y to your sdkconfig
+//!
+//! ```no_run
+//! use esp_idf_hal::peripherals::Peripherals;
+//! use esp_idf_svc::eventloop::EspSystemEventLoop;
+//! use esp_idf_svc::nvs::EspDefaultNvsPartition;
+//! use esp_idf_svc::wifi::EspWifi;
+//! use esp_idf_svc::wifi_dpp::EspWifiDpp;
+//!
+//! let peripherals = Peripherals::take().unwrap();
+//! let sysloop = EspSystemEventLoop::take()?;
+//! let nvs = EspDefaultNvsPartition::take()?;
+//! let mut wifi = EspWifi::new(peripherals.modem, sysloop, Some(nvs))?;
+//!
+//! let channels = [6];
+//! // Test key, please use secure keys for your project (or None to generate one on the fly)!
+//! let privkey = Some([
+//!     0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff,
+//!     0x00, 0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88, 0x99, 0xaa, 0xbb, 0xcc, 0xdd, 0xee, 0xff,
+//! ]);
+//! let associated_data = None;
+//! loop {
+//!     let dpp = EspWifiDpp::generate_qrcode(
+//!         &mut wifi,
+//!         &channels,
+//!         privkey.as_ref(),
+//!         associated_data)?;
+//!     log::info!("Got QR code text: {}", dpp.get_bootstrapped_data().0);
+//!
+//!     match dpp.start_listen()?.wait_for_credentials() {
+//!         Ok(c) => break,
+//!         Err(e) => {
+//!             // Will generate the same QR code again since the inputs have not changed...
+//!             log::error!("DPP error: {e}, bootstrapping again...");
+//!         }
+//!     }
+//! }
+//! ```
+
+use alloc::sync::Arc;
+use alloc::sync::Weak;
+use core::fmt::Write;
+use core::marker::PhantomData;
+use core::mem;
+use core::ops::{Deref, DerefMut};
+use core::ptr;
+use core::time::Duration;
+
+use ::log::*;
+use embedded_svc::wifi::{ClientConfiguration, Configuration};
+use esp_idf_sys::EspError;
+use esp_idf_sys::*;
+
+use crate::private::common::Newtype;
+use crate::private::cstr::*;
+use crate::private::mutex::{Mutex, RawMutex};
+use crate::private::waitable::Waitable;
+use crate::wifi::EspWifi;
+
+/// Global singleton that proves we can't have imbalanced access to esp_supp_dpp_init/deinit.
+/// This is statically enforced through requiring a mutable borrow of EspWifi in the API.
+struct DppInitialized {
+    /// Holds the most recently received callback event that is yet to be processed/handled.
+    pending: Waitable<Option<DppState>>,
+}
+
+/// Global weak reference so that we can respond to the stateless C callbacks provided in
+/// esp_supp_dpp_init.
+static DPP_INITIALIZED: Mutex<Option<Weak<DppInitialized>>> = Mutex::wrap(RawMutex::new(), None);
+
+impl DppInitialized {
```

Comment:
I guess I'm missing a detail here, but I'm a bit lost as to _why_ we need `DppInitialized` in the first place? Here's my thinking:
* `EspWifi` is a singleton; you cannot have two instances of it, alive at any time
* Ergo, there can be *only one instance, ever* of `EspWifiDpp` alive at any time, as by design, creating such an instance needs a mutable borrow of `EspWifi`
* You initialize DPP when constructing `EspWifiDpp`
* You de-initialize DPP when dropping `EspWifiDpp`

What am I missing?

Reply:
I can combine them, but the main reason I kept them separate was so that it was very clear the global singleton (DPP_INITIALIZED) was literally tied to the esp_supp_dpp_init/deinit lifecycle, whereas EspWifiDpp was tied to the _bootstrapped_ lifecycle after you call esp_supp_dpp_bootstrap_gen.  In theory, you could bootstrap multiple times without deinit which would require fewer changes with the way it's currently designed than it would if DppInitialized/EspWifiDpp were combined.  I'll leave it up to you if you prefer to reduce the total amount of code further in exchange for a maybe heavier refactoring needed later if we want to add features (which I suspect will happen since Espressif's support for DPP is pretty lacking at the moment).